Patch Ctrl-R to immediately run the command after selection
Bind Ctrl-F to the old Ctrl-R

@@ -106,7 +106,7 @@
 fi
 
 # CTRL-R - Paste the selected command from history into the command line
-fzf-history-widget() {
+__fzf-history() {
   local selected
   setopt localoptions noglobsubst noposixbuiltins pipefail no_aliases noglob nobash_rematch 2> /dev/null
   # Ensure the associative history array, which maps event numbers to the full
@@ -116,11 +116,32 @@
       perl -0 -ne 'if (!$seen{(/^\s*[0-9]+\**\t(.*)/s, $1)}++) { s/\n/\n\t/g; print; }' |
       FZF_DEFAULT_OPTS=$(__fzf_defaults "" "-n2..,.. --scheme=history --bind=ctrl-r:toggle-sort --wrap-sign '\t↳ ' --highlight-line ${FZF_CTRL_R_OPTS-} --query=${(qqq)LBUFFER} +m --read0") \
       FZF_DEFAULT_OPTS_FILE='' $(__fzfcmd))"
+    echo $selected
   else
     selected="$(fc -rl 1 | awk '{ cmd=$0; sub(/^[ \t]*[0-9]+\**[ \t]+/, "", cmd); if (!seen[cmd]++) print $0 }' |
       FZF_DEFAULT_OPTS=$(__fzf_defaults "" "-n2..,.. --scheme=history --bind=ctrl-r:toggle-sort --wrap-sign '\t↳ ' --highlight-line ${FZF_CTRL_R_OPTS-} --query=${(qqq)LBUFFER} +m") \
       FZF_DEFAULT_OPTS_FILE='' $(__fzfcmd))"
+    echo $selected
+  fi
+}
+fzf-history-widget() {
+  local selected
+  selected=$(__fzf-history)
+  local ret=$?
+  if [ -n "$selected" ]; then
+    if [[ $(awk '{print $1; exit}' <<< "$selected") =~ ^[1-9][0-9]* ]]; then
+      zle vi-fetch-history -n $MATCH
+    else # selected is a custom query, not from history
+      LBUFFER="$selected"
+    fi
+    zle accept-line
   fi
+  zle reset-prompt
+  return $ret
+}
+fzf-history-widget2() {
+  local selected
+  selected=$(__fzf-history)
   local ret=$?
   if [ -n "$selected" ]; then
     if [[ $(awk '{print $1; exit}' <<< "$selected") =~ ^[1-9][0-9]* ]]; then
@@ -136,6 +157,10 @@
 bindkey -M emacs '^R' fzf-history-widget
 bindkey -M vicmd '^R' fzf-history-widget
 bindkey -M viins '^R' fzf-history-widget
+zle     -N            fzf-history-widget2
+bindkey -M emacs '^F' fzf-history-widget2
+bindkey -M vicmd '^F' fzf-history-widget2
+bindkey -M viins '^F' fzf-history-widget2
 fi
 
 } always {
